c_src/membrane_vpx_plugin/vpx_decoder.c

```diff
@@ -0,0 +1,150 @@
+#include "vpx_decoder.h"
+
+void handle_destroy_state(UnifexEnv *env, State *state) {
+  UNIFEX_UNUSED(env);
+
+  vpx_codec_destroy(&state->codec_context);
+}
+
+UNIFEX_TERM create(UnifexEnv *env, Codec codec) {
+  UNIFEX_TERM result;
+  State *state = unifex_alloc_state(env);
+
+  switch (codec) {
+  case CODEC_VP8:
+    state->codec_interface = vpx_codec_vp8_dx();
+    break;
+  case CODEC_VP9:
+    state->codec_interface = vpx_codec_vp9_dx();
+    break;
+  }
+
+  if (vpx_codec_dec_init(&state->codec_context, state->codec_interface, NULL,
+                         0)) {
+    result = create_result_error(env, "Failed to initialize decoder");
+    unifex_release_state(env, state);
+    return result;
+  }
+  result = create_result_ok(env, state);
+  unifex_release_state(env, state);
+  return result;
+}
+
+Dimensions get_plane_dimensions(const vpx_image_t *img, int plane) {
+  const int height = (plane > 0 && img->y_chroma_shift > 0)
+                         ? (img->d_h + 1) >> img->y_chroma_shift
+                         : img->d_h;
+
+  int width = (plane > 0 && img->x_chroma_shift > 0)
+                  ? (img->d_w + 1) >> img->x_chroma_shift
+                  : img->d_w;
+
+  // Fixing NV12 chroma width if it is odd
+  if (img->fmt == VPX_IMG_FMT_NV12 && plane == 1)
+    width = (width + 1) & ~1;
+
+  return (Dimensions){width, height};
+}
+size_t get_image_byte_size(const vpx_image_t *img) {
+  const int bytes_per_pixel = (img->fmt & VPX_IMG_FMT_HIGHBITDEPTH) ? 2 : 1;
+  const int number_of_planes = (img->fmt == VPX_IMG_FMT_NV12) ? 2 : 3;
+
+  size_t image_size = 0;
+
+  for (int plane = 0; plane < number_of_planes; ++plane) {
+    Dimensions plane_dimensions = get_plane_dimensions(img, plane);
+    image_size +=
+        plane_dimensions.width * plane_dimensions.height * bytes_per_pixel;
+  }
+  return image_size;
+}
+
+void get_output_frame_from_image(const vpx_image_t *img,
+                                 UnifexPayload *output_frame) {
+  const int bytes_per_pixel = (img->fmt & VPX_IMG_FMT_HIGHBITDEPTH) ? 2 : 1;
+
+  // Assuming that for nv12 we write all chroma data at once
+  const int number_of_planes = (img->fmt == VPX_IMG_FMT_NV12) ? 2 : 3;
```

Comment:
could we somehow unify it with `const int bytes_per_pixel` and `const int number_of_planes` defined in `get_image_byte_size`?

Reply:
i could abstract this to two functions, but i don't think this would make this more readable, since these are simple one-liners